lib/core/CJsonStateRestoreTraverser.cc

```diff
@@ -240,19 +240,21 @@ bool CJsonStateRestoreTraverser::parseNext(bool remember) {
 
         m_Handler.s_RememberValue = remember;
 
-        char c = '\0';
-        m_ReadStream.get(c);
-        if (c == '\0') {
-            break;
+        if (m_BytesRemaining == 0) {
+            m_BytesRemaining = BUFFER_SIZE - 1;
+            ::memset(m_Buffer, '\0', BUFFER_SIZE);
+            m_ReadStream.get(m_Buffer, BUFFER_SIZE, '\0');
+            m_BufferPtr = m_Buffer;
         }
 
         json::error_code ec;
-        m_Reader.write_some(true, &c, 1, ec);
+        std::size_t written = m_Reader.write_some(true, m_BufferPtr++, 1, ec);
```

Comment:
Does this work?

```suggestion
        std::size_t written = m_Reader.write_some(m_ReadStream.eof() == false, m_BufferPtr, m_BytesRemaining, ec);
        m_BufferPtr += written;
```

If it does it will be much more efficient.

Reply:
Sadly no. This results in `parseNext` not returning for each complete JSON token that has been processed. In this respect the `boost::json` SAX parser is different from the `RapidJSON` parser that supplies the `IterativeParseNext` function that parses one single JSON token from its input.

Reply:
Oh yes, I see. Multiple callbacks can get called but they'll just trample each other's changes _before_ `parseNext` returns.

But then it seems doubly important to keep `m_BufferPtr` and `m_RemainingBytes` perfectly in sync. So we should have `m_BufferPtr += written` and `m_RemainingBytes -= written` as adjacent lines immediately after the `write_some` call, and not use `m_BufferPtr++` before we know the result of `write_some`.